packages/tpetra/core/src/Tpetra_Details_residual.hpp

```diff
@@ -0,0 +1,294 @@
+      (X.getNumVectors () != R.getNumVectors (), std::runtime_error,
+       "X.getNumVectors() = " << X.getNumVectors () << " != "
+       "R.getNumVectors() = " << R.getNumVectors () << ".");
+
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (X.getLocalLength () !=
+       A.getColMap ()->getNodeNumElements (), std::runtime_error,
+       "X has the wrong number of local rows.  "
+       "X.getLocalLength() = " << X.getLocalLength () << " != "
+       "A.getColMap()->getNodeNumElements() = " <<
+       A.getColMap ()->getNodeNumElements () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (R.getLocalLength () !=
+       A.getRowMap ()->getNodeNumElements (), std::runtime_error,
+       "R has the wrong number of local rows.  "
+       "R.getLocalLength() = " << R.getLocalLength () << " != "
+       "A.getRowMap()->getNodeNumElements() = " <<
+       A.getRowMap ()->getNodeNumElements () << ".");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (B.getLocalLength () !=
+       A.getRowMap ()->getNodeNumElements (), std::runtime_error,
+       "B has the wrong number of local rows.  "
+       "B.getLocalLength() = " << B.getLocalLength () << " != "
+       "A.getRowMap()->getNodeNumElements() = " <<
+       A.getRowMap ()->getNodeNumElements () << ".");
+
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (! A.isFillComplete (), std::runtime_error, "The matrix A is not "
+       "fill complete.  You must call fillComplete() (possibly with "
+       "domain and range Map arguments) without an intervening "
+       "resumeFill() call before you may call this method.");
+    TEUCHOS_TEST_FOR_EXCEPTION
+      (! X.isConstantStride () || ! R.isConstantStride () || ! B.isConstantStride (),
+       std::runtime_error, "X, Y and B must be constant stride.");
+    // If the two pointers are NULL, then they don't alias one
+    // another, even though they are equal.
+    TEUCHOS_TEST_FOR_EXCEPTION
+      ((X_lcl.data () == R_lcl.data () && X_lcl.data () != nullptr) ||
+       (X_lcl.data () == B_lcl.data () && X_lcl.data () != nullptr),
+       std::runtime_error, "X, Y and R may not alias one another.");
+  }
+      
+  // This is currently a "reference implementation" waiting until Kokkos Kernels provides
+  // a residual kernel.
+  SC one = Teuchos::ScalarTraits<SC>::one(), negone = -one, zero = Teuchos::ScalarTraits<SC>::zero();    
+  A.localApply(X,R,Teuchos::NO_TRANS, one, zero);
+  R.update(one,B,negone);
+}
+    
+
+//! Computes R = B - A * X 
+template<class SC, class LO, class GO, class NO>
+void residual(const Operator<SC,LO,GO,NO> &   Aop,
+              const MultiVector<SC,LO,GO,NO> & X_in,
+              const MultiVector<SC,LO,GO,NO> & B_in,
+              MultiVector<SC,LO,GO,NO> & R_in) {
+  using Tpetra::Details::ProfilingRegion;
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  using Teuchos::rcp_const_cast;
+  using Teuchos::rcpFromRef;
+  
+  const CrsMatrix<SC,LO,GO,NO> * Apt  = dynamic_cast<const CrsMatrix<SC,LO,GO,NO>*>(&Aop);
+  if(!Apt) {
```

Comment:
You could just call the `CrsMatrix` overload here.  That means you'll only have to optimize one implementation.